pwndbg/heap/ptmalloc.py

```diff
@@ -1191,14 +1191,55 @@ def bin_at(self, index, arena_addr=None):
         chain_fd = get_chain(front, fd_offset)
         chain_bk = get_chain(back, bk_offset)
 
-        # check if bin[index] points to itself (is empty)
-        if len(chain_fd) == len(chain_bk) == 2 and chain_fd[0] == chain_bk[0]:
-            chain_fd = [0]
-            chain_bk = [0]
-
-        # check if corrupted
-        elif chain_fd[:-1] != chain_bk[:-2][::-1] + [chain_bk[-2]]:
+        if len(chain_fd) != len(chain_bk):
+            # If the chain lengths aren't equal, the chain is corrupted
             is_chain_corrupted = True
+        elif len(chain_fd) < 2 or len(chain_bk) < 2:
+            # Chains containing less than two entries are corrupted, as the smallest
+            # chain (an empty bin) would look like something like `[main_arena+88, 0]`.
+            is_corrupted = True
+        elif len(chain_fd) == len(chain_bk) == 2 and chain_fd == chain_bk and chain_fd[-1] == 0:
+            # Check if bin[index] points to itself (is empty)
+
+            # If the fd pointer is in libc data section, we can be more
+            # confident that this list is actually empty and not just
+            # corrupted. If we can't find any mapping for the pointer or it's in
+            # some other mapping, we assume the bin is corrupted
+            page = pwndbg.gdblib.vmmap.find(chain_fd[0])
+            if page and page.rw and "libc" in page.objfile:
```

Comment:
I do not very understand the logic behind this yet, so probably I misunderstood something:

I have some concerns about the check here:
1. What if vmmap does not work well(e.g. for the native qemu-user), will it make pwndbg think the empty bins are corrupted bins?
2. If we are debugging a binary with statically linked libc, will it make pwndbg think the empty bins are corrupted bins since `"libc" in page.objfile` is never be `True`?
3. IIRC, bins in non-main arena might not in someone's .data sectioin, does this check also true for the bins not in main_arena?
4. Seems like it might pass the check if the address is on the same page as the .data section, but not actually "in" the .data section? Maybe parsing `info files` will be better if we want to know if the address is in .data section or not?

Reply:
> What if vmmap does not work well(e.g. for the native qemu-user), will it make pwndbg think the empty bins are corrupted bins?

Hmm, I think you're right. Maybe we should have a function that returns whether the `vmmap` output can be trusted or not?

> If we are debugging a binary with statically linked libc, will it make pwndbg think the empty bins are corrupted bins since "libc" in page.objfile is never be True?

I didn't think about this, you're right.

> IIRC, bins in non-main arena might not in someone's .data sectioin, does this check also true for the bins not in main_arena?

Good point. I can add a check here to only do this logic if it's the main arena?

> Seems like it might pass the check if the address is on the same page as the .data section, but not actually "in" the .data section? Maybe parsing info files will be better if we want to know if the address is in .data section or not?

Sure, I'm more worried though about a non-corrupted chain being marked as corrupted, not a corrupted chain being marked as not corrupted. The latter is just best effort on our part, while the former would be considered a bug.

Reply:
> I can add a check here to only do this logic if it's the main arena?

Hmm, instead of still checking the fd_chain/bk_chain point to the .data section or not, could we check the front and back pointers instead of the fd_chain/bk_chain?
I mean, something like this:
```diff
--- a/pwndbg/heap/ptmalloc.py
+++ b/pwndbg/heap/ptmalloc.py
@@ -1200,13 +1200,7 @@ class GlibcMemoryAllocator(pwndbg.heap.heap.MemoryAllocator):
             is_corrupted = True
         elif len(chain_fd) == len(chain_bk) == 2 and chain_fd == chain_bk and chain_fd[-1] == 0:
             # Check if bin[index] points to itself (is empty)
-
-            # If the fd pointer is in libc data section, we can be more
-            # confident that this list is actually empty and not just
-            # corrupted. If we can't find any mapping for the pointer or it's in
-            # some other mapping, we assume the bin is corrupted
-            page = pwndbg.gdblib.vmmap.find(chain_fd[0])
-            if page and page.rw and "libc" in page.objfile:
+            if front == back and front == current_base:
                 chain_fd = [0]
                 chain_bk = [0]
             else:
```
(This is entirely from my observation, I'm not very sure :p)